### PR TITLE
Disable/Enable logs by name

### PR DESCRIPTION
**Pull request by proflayton**
Expose an api to allow disabling of logs my the msg name

### CODE REVIEW

File: larch.js [hunk 1]
```diff
@@ -63,6 +65,14 @@ function enableDebug(enabled) {
     }
 };
 
+Larch.prototype.disableLog = function disableLog(name) {
+    this.disabledLogs[name] = true;
+};
+
+Larch.prototype.enableLog = function enableLog(name) {
+    delete this.disabledLogs[name];
```
Comment (Raynos) [inline larch.js]: nitpick: `this.disabledLogs[name] = false`

File: larch.js [hunk 2]
```diff
@@ -168,60 +178,74 @@ Larch.prototype.destroy = function destroy(cb) {
     }
 };
 
+Larch.prototype.logIsEnabled = function logIsEnabled(msg) {
```
Comment (Raynos) [inline larch.js]: `level` or `msg` ?

File: larch.js [hunk 3]
```diff
@@ -168,60 +178,74 @@ Larch.prototype.destroy = function destroy(cb) {
...
+            return cb();
+        }
+        return;
+    }
+    this[fn](level, msg, meta, cb);
+};
+
 Larch.prototype.trace = function trace(msg, meta, cb) {
-    this.log('trace', msg, meta, cb);
+    this.tryLog('log', 'trace', msg, meta, cb);
```
Comment (Raynos) [inline larch.js]: let's implement `tryLog` and `stryLog`, i suspect its faster.

File: larch.js [hunk 4]
```diff
@@ -168,60 +178,74 @@ Larch.prototype.destroy = function destroy(cb) {
...
+};
+
+Larch.prototype.tryLog = function tryLog(fn, level, msg, meta, cb) {
+    if (!this.logIsEnabled(msg)) {
+        if (typeof cb === 'function') {
+            return cb();
+        }
+        return;
+    }
+    this[fn](level, msg, meta, cb);
```
Comment (Raynos) [inline larch.js]: dynamic method lookup is #sad

File: larch.js [hunk 5]
```diff
@@ -168,60 +178,74 @@ Larch.prototype.destroy = function destroy(cb) {
...
 };
 
+Larch.prototype.logIsEnabled = function logIsEnabled(msg) {
+    return !this.disabledLogs[msg];
+};
+
+Larch.prototype.tryLog = function tryLog(fn, level, msg, meta, cb) {
+    if (!this.logIsEnabled(msg)) {
+        if (typeof cb === 'function') {
+            return cb();
```
Comment (Raynos) [inline larch.js]: nitpick: no need for `return` ?

File: larch.js (hunk 2 above)
Comment (proflayton) [inline larch.js]: I should call this `name`

File: larch.js (hunk 3 above)
Comment (proflayton) [inline larch.js]: youre probably right. I just hate code duplication :P

File: larch.js [hunk 6]
```diff
@@ -39,6 +39,8 @@ function Larch(options) {
     assert(Array.isArray(self.backends), 'options.backends must be array');
     self.enableDebug(true);
 
+    self.disabledLogs = {};
```
Comment (Raynos) [inline larch.js]: nitpick: `self.disabledLogs = Object.create(null)`

Comment (proflayton) [inline larch.js]: Huh I dont think I have seen this pattern. What is the difference?

Comment (Raynos) [inline larch.js]: http://www.2ality.com/2013/10/dict-pattern.html

`!!disabledLogs['hasOwnProperty'] // true`

File: larch.js [hunk 7]
```diff
@@ -168,60 +178,84 @@ Larch.prototype.destroy = function destroy(cb) {
...
 };
 
 Larch.prototype.access = function access(msg, meta, cb) {
-    this.log('access', msg, meta, cb);
+    this.tryLog('access', msg, meta, cb);
 };
 
 Larch.prototype.warn = function warn(msg, meta, cb) {
-    this.log('warn', msg, meta, cb);
+    this.tryLog('log', 'warn', msg, meta, cb);
```
Comment (Raynos) [inline larch.js]: warn

File: larch.js (hunk 6 above)
Comment (proflayton) [inline larch.js]: I see. That is super good to know

File: larch.js (hunk 7 above)
Comment (proflayton) [inline larch.js]: o.o thanks

**Comment by Raynos** [comment]
lgtm.

**Pull request merged**